lib/charms/loki_k8s/v0/charm_logging.py

```diff
@@ -0,0 +1,463 @@
+        cert: Optional[str] = None,
+    ):
+        """Create new Loki logging handler.
+
+        Arguments:
+            url: Endpoint used to send log entries to Loki (e.g. `https://my-loki-instance/loki/api/v1/push`).
+            tags: Default tags added to every log record.
+
+            # FIXME: Session expects a .pem file it says
+            cert: Optional absolute path to cert file for TLS auth.
+
+        """
+        super().__init__()
+        self.emitter = LokiEmitter(url, tags, cert)
+
+    def handleError(self, record):  # noqa: N802
+        """Close emitter and let default handler take actions on error."""
+        self.emitter.close()
+        super().handleError(record)
+
+    def emit(self, record: logging.LogRecord):
+        """Send log record to Loki."""
+        # noinspection PyBroadException
+        try:
+            self.emitter(record, self.format(record))
+        except Exception:
+            self.handleError(record)
+
+
+def is_enabled() -> bool:
+    """Whether charm logging is enabled."""
+    return os.getenv(CHARM_LOGGING_ENABLED, "1") == "1"
+
+
+class CharmLoggingError(Exception):
+    """Base class for all exceptions raised by this module."""
+
+
+class InvalidEndpointError(CharmLoggingError):
+    """Raised if an endpoint is invalid."""
+
+
+@contextmanager
+def charm_logging_disabled():
+    """Contextmanager to temporarily disable charm logging.
+
+    For usage in tests.
+    """
+    previous = os.getenv(CHARM_LOGGING_ENABLED, "1")
+    os.environ[CHARM_LOGGING_ENABLED] = "0"
+    yield
+    os.environ[CHARM_LOGGING_ENABLED] = previous
+
+
+_C = TypeVar("_C", bound=Type[CharmBase])
+_T = TypeVar("_T", bound=type)
+_F = TypeVar("_F", bound=Type[Callable])
+
+
+def _get_logging_endpoints(logging_endpoints_getter, self, charm):
+    if isinstance(logging_endpoints_getter, property):
+        logging_endpoints = logging_endpoints_getter.__get__(self)
+    else:  # method or callable
+        logging_endpoints = logging_endpoints_getter(self)
+
+    if logging_endpoints is None:
+        logger.debug(
+            f"{charm}.{logging_endpoints_getter} returned None; quietly disabling "
+            f"charm_logging for the run."
+        )
+        return None
+
+    errors = []
+    logging_endpoints = tuple(logging_endpoints)
+    sanitized_logging_endponts = []
+    for endpoint in logging_endpoints:
+        if isinstance(endpoint, str):
+            sanitized_logging_endponts.append(endpoint)
+        else:
+            errors.append(endpoint)
+
+    if errors:
+        if sanitized_logging_endponts:
+            logger.error(
+                f"{charm}.{logging_endpoints_getter} returned some invalid endpoint strings: {errors}"
+            )
+        else:
+            logger.error(
+                f"{charm}.{logging_endpoints_getter} should return an iterable of Loki push-api (compatible) endpoints (strings); "
+                f"got {errors} instead."
+            )
```

Comment:
This way is clearer and shorter:


```suggestion
        message = f"{charm}.{logging_endpoints_getter} should return an iterable of Loki push-api (compatible) endpoints (strings); got {errors} instead."

        if sanitized_logging_endponts:
            message = f"{charm}.{logging_endpoints_getter} returned some invalid endpoint strings: {errors}"

        logger.error(message)
```

Reply:
cleaned it up, thanks for the tip